File: Talleres/Taller 2/code/BinaryInversedIterative.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Convert an integer number into binary list
string DecimalToBinary(int n)
{
    string binary = "";
    int i = 0;
    while (n > 0)
    {
        char c = n % 2 + '0';
        binary += c;
        n = n / 2;
        i++;
    }
    return binary;
}

// Invert a chain given a list begin and end
string InvertChainInt(string S)
{
    if (S.size() == 1)
    {
        string aux = "";
        aux += S[0];
        return aux;
    }
    string result = "";
    for (int i = S.size() - 1; i >= 0; i--)
    {
        result += S[i];
    }
    return result;
}

int BinaryToDecimal(string S)
{
    int decimal = 0;
    int i = 0;
    string aux = InvertChainInt(S);
    for (int i = 0; i < aux.size(); i++)
    {
        int c = aux[i] - '0';
        int p = c * pow(2, aux.size() - i - 1);
        decimal += p;
    }
    return decimal;
}

File: Talleres/Taller 2/code/python/BinaryInversedIterative.py
from math import pow
#Convert an integer number into binary list
def DecimalToBinary(n):
    binary=[]
    while n !=0:
        binary.insert(0,str(n%2))
        n //= 2
    return binary

#Invert a chain given a list begin and end
def InvertChain(S):
    if len(S)==1:
        return S[0]
    result  = []
    for i in range (len(S)-1,-1,-1):
        result.append(S[i])
    return result

#Convert a list of binary into an integer
def BinaryToDecimal(numero_binario):
    decimal = 0
    i=0
    inverted = InvertChain(numero_binario)
    for i in range(len(inverted)):
        p = int(pow(int(inverted[i])*2,i))
        decimal = decimal + p
    return decimal

print("Binary Inversed Iterative")
n = 345
binary= DecimalToBinary(n)
print("Original:", n, binary)
binarioInverso = InvertChain(binary)
decimal= BinaryToDecimal(binarioInverso)
print("Inverso:", decimal, binarioInverso)
